sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -218,6 +228,54 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
     result.copy(expectedOutputAttributes = Some(relation.output))
   }
 
+  private def inferIfNeeded(
+      relation: CatalogRelation,
+      options: Map[String, String],
+      fileFormat: FileFormat,
+      fileIndexOpt: Option[FileIndex] = None): (StructType, CatalogTable) = {
+    val inferenceMode = sparkSession.sessionState.conf.caseSensitiveInferenceMode
+    val shouldInfer = (inferenceMode != NEVER_INFER) && !relation.tableMeta.schemaPreservesCase
+    val tableName = relation.tableMeta.identifier.unquotedString
+    if (shouldInfer) {
+      logInfo(s"Inferring case-sensitive schema for table $tableName (inference mode: " +
+        s"$inferenceMode)")
+      val fileIndex = fileIndexOpt.getOrElse {
+        val rootPath = new Path(relation.tableMeta.location)
+        new InMemoryFileIndex(sparkSession, Seq(rootPath), options, None)
+      }
+
+      val inferredSchema = fileFormat
+        .inferSchema(
+          sparkSession,
+          options,
+          fileIndex.listFiles(Nil).flatMap(_.files))
```

Comment:
@budde @cloud-fan  just wondering if there was a reason we didn't provide a partition filter here instead of `Nil`? I'm seeing a case where a query, with a partition predicate specified, is hitting this piece of code and all partitions are being listed unnecessarily. I think it can be solved by setting the inference mode to `NEVER_INFER` but not sure if that is ideal or can have any side effects.

Reply:
schema inference needs to look at the entire data set.

Reply:
To expand on this, let's say you are dealing with a dataset that has an optional field in its schema. If none of the data files for the partition(s) you filter by contain this field then it will never be returned when using schema inference. This may not matter in your particular usecase but IMO inspecting all of the files when inferring the schema is the safer approach in the general case. I believe you can configure a threshold for the number of data files at which the schema inference step will be distributed across the cluster and performed as a map reduce job vs. running on the driver node.

If the full schema is already known beforehand and you're not using any sort of metastore to track table/schema state, using `NEVER_INFER` in conjunction with manually providing the schema as a read option might be a good approach.